### ISSUE TITLE
Decimal type does not allow to store some ISO20022 fields

### ISSUE THREAD

**Issue by filmackay** [digital-asset/daml#1706]
We're having a hard time with differences in `Decimal` between DAML and other environments. In particular the usage of it in ISO20022 (yes - a big subject area, but stick with me here..). ISO20022 uses decimals in many contexts, but they are specializations of the most general xsd definition for the type `DecimalNumber`:
```
    <xs:simpleType name="DecimalNumber">
        <xs:restriction base="xs:decimal">
            <xs:fractionDigits value="17"/>
            <xs:totalDigits value="18"/>
        </xs:restriction>
    </xs:simpleType>
```
Example: https://www.iso20022.org/sites/default/files/documents/messages/sese/schemas/sese.023.001.09.xsd

- How do we deal with scenarios that need > 10 decimal places
- Less bounded decimals / dependent types like Decimal[28,8]? :)
- Should we consider using Sci notation instead
- Note that the above XSD notation is different from SQL, in that the latter has a fixed fractional digits

**Comment by neil-da** [comment]
There are really two options: 1) the runtime team gives you more power. 2) you abstract over this with a newtype and your own number representation. I think you have all the tools to do option 2, but it's going to be a lot less efficient and store horribly on the ledger for nanobots etc.

**Comment by bitonic** [comment]
@filmackay thanks a lot for pointing this out.

I am a bit confused by the schema you quote. `<xs:totalDigits value="18"/>` seems to imply that the total number of digits is 18. At the same time `<xs:fractionDigits value="17"/>` seems to say that there are 17 digits after the dot.

Does this mean that ISO20022 requires numbers to have at most one integral digits and at most 17 fractional digits?

In any case, I think the right thing to do here is to allow users to use numbers with more than 10 decimal places easily. We have many options here, I'll think a bit harder on what the best one is...

**Comment by raphael-speyer-da** [comment]
IIUC the schema means that their `DecimalNumber` has a max precision of 18, and max scale of 17, so we'd need to support.

```
[
  123456789012345678.0,
  0.12345678901234567,
  1.23456789012345678
] : [Decimal]
```

**Comment by dsun-da** [comment]
Please note the decimal type over ledger API:
as I have encountered the following error when testing a decimal with 11 FD (fractional digits)
```
Submit to ledger failed. Status: INVALID_ARGUMENT : Invalid argument: Could not read Decimal string "0.00000000001"  
```

**Comment by filmackay** [comment]
> I am a bit confused by the schema you quote. `<xs:totalDigits value="18"/>` seems to imply that the total number of digits is 18. At the same time `<xs:fractionDigits value="17"/>` seems to say that there are 17 digits after the dot.
The schema-squared is:
```<xs:totalDigits
           fixed          = xs:boolean : “false”
           id             = xs:ID
           value          = xs:positiveInteger
           {any attributes with non-schema namespace}
           >
           Content: (xs:annotation?)
</xs:totalDigits>```
Where `fixed` is not supplied (in this case) it is a "maximum" value. Note that leading & trailing zeros are removed.
Similarly `fractionDigits` as also a "maximum" in this case:
```<xs:fractionDigits
           fixed          = xs:boolean : “false”
           id             = xs:ID
           value          = xs:nonNegativeInteger
           {any attributes with non-schema namespace}
           >
           Content: (xs:annotation?)
</xs:fractionDigits>```
> In any case, I think the right thing to do here is to allow users to use numbers with more than 10 decimal places easily. We have many options here, I'll think a bit harder on what the best one is...
I think the problem is that SQL & XSD make you pin the precision/size of the decimal, in dependent-type-like form - whereas `Decimal` in DAML is completely dynamic. You need a very wide decimal to encompass all possible ISO/XSD-encodable values; and additionally to face the problem of how to put this into a SQL back-end where you need to pin the precision/size for persistence. Covering ISO20022 should be fine, but the best you could store in SQL would be `DECIMAL(28,17)` meaning trouble storing the value `100,000,000,000`.

**Comment by filmackay** [comment]
> Please note the decimal type over ledger API:
> as I have encountered the following error when testing a decimal with 11 FD (fractional digits)
> 
> ```
> Submit to ledger failed. Status: INVALID_ARGUMENT : Invalid argument: Could not read Decimal string "0.00000000001"  
> ```

Yes ATM this is by design, according to DAML-LF spec.

**Comment by filmackay** [comment]
> There are really two options: 1) the runtime team gives you more power. 2) you abstract over this with a newtype and your own number representation. I think you have all the tools to do option 2, but it's going to be a lot less efficient and store horribly on the ledger for nanobots etc.

@neil-da yes I had wondered about 2) a in-DAML BigDecimal implementation - we could interop this with codegen but the SQL persistence would be a bit unexpected!

**Comment by gerolf-da** [comment]
This has been fixed by #2289.
